src/Queue/DoctrineQueue.php

```diff
@@ -182,19 +225,162 @@ public function pop(array $options = []): ?JobInterface
                 }
             }
 
-            $conn->commit();
+            $this->connection->commit();
         } catch (DBALException $e) {
-            $conn->rollback();
-            $conn->close();
+            $this->connection->rollback();
+            $this->connection->close();
             throw new RuntimeException($e->getMessage(), $e->getCode(), $e);
         }
 
         if ($row === false) {
             return null;
         }
 
-        // Add job ID to meta data
-        return $this->unserializeJob($row['data'], ['__id__' => $row['id']]);
+        return $row;
+    }
+
+    /**
+     * This generic variant of pop() works like this:
+     * - UPDATE query (limit 1) that sets the job into a new status (STATUS_ALLOCATING) with a unique (!) worker ID
+     * - If the row count of the UPDATE is zero -> no Jobs are ready for execution -> return null.
+     * - If the row count of the UPDATE is greater 1 -> too many Jobs were reserved -> rollback and throw an exception. This should not happen!
+     * - If the row count of the UPDATE is 1 -> we've marked a job for us
+     * - SELECT the job via the unique identifier
+     * - UPDATE it to STATUS_RUNNING
+     * - Return its data.
+     *
+     * It uses 3 queries (UPDATE, SELECT, UPDATE) to pop a job from the queue table without any locking or transaction.
+     *
+     * @param array $options
+     *
+     * @return array|null
+     *
+     * @throws \Exception
+     */
+    protected function popMysql(array $options = []): ?array
+    {
+        $time      = microtime(true);
+        $micro     = sprintf("%06d", ($time - floor($time)) * 1000000);
+        $this->now = new DateTime(
+            date('Y-m-d H:i:s.' . $micro, $time),
+            new DateTimeZone(date_default_timezone_get())
+        );
```

Comment:
Only if you feel like it: you could extract this in a separate function.

Reply:
Dont know why this function is needed at all.
`new DateTime()` already sets the time to NOW including microseconds. See https://3v4l.org/tvahg

Reply:
Cause these versions got it wrong

https://3v4l.org/tvahg#v550

Reply:
Oh! Interesting.
PHP 7.1.4+ seems to set the ms part correctly and the composer.json has a "^7.2" constraint on PHP. That means we dont need it anymore, correct?

Reply:
Correct